hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacySqlAstTranslator.java

```diff
@@ -0,0 +1,213 @@
+			this.renderAsArray = true;
+			inSubQueryPredicate.getSubQuery().accept( this );
+			this.renderAsArray = renderAsArray;
+		}
+		else {
+			super.visitInSubQueryPredicate( inSubQueryPredicate );
+		}
+	}
+
+	@Override
+	protected void visitSqlSelections(SelectClause selectClause) {
+		final boolean renderAsArray = this.renderAsArray;
+		this.renderAsArray = false;
+		if ( renderAsArray ) {
+			append( OPEN_PARENTHESIS );
+		}
+		super.visitSqlSelections( selectClause );
+		if ( renderAsArray ) {
+			append( CLOSE_PARENTHESIS );
+		}
+	}
+
+	@Override
+	protected void renderPartitionItem(Expression expression) {
+		if ( expression instanceof Literal ) {
+			appendSql( "'0' || '0'" );
+		}
+		else if ( expression instanceof Summarization ) {
+			// This could theoretically be emulated by rendering all grouping variations of the query and
+			// connect them via union all but that's probably pretty inefficient and would have to happen
+			// on the query spec level
+			throw new UnsupportedOperationException( "Summarization is not supported by DBMS!" );
+		}
+		else {
+			expression.accept( this );
+		}
+	}
+
+	@Override
+	public void visitBinaryArithmeticExpression(BinaryArithmeticExpression arithmeticExpression) {
+		appendSql( OPEN_PARENTHESIS );
+		render( arithmeticExpression.getLeftHandOperand(), SqlAstNodeRenderingMode.NO_PLAIN_PARAMETER );
+		appendSql( arithmeticExpression.getOperator().getOperatorSqlTextString() );
+		render( arithmeticExpression.getRightHandOperand(), SqlAstNodeRenderingMode.NO_PLAIN_PARAMETER );
+		appendSql( CLOSE_PARENTHESIS );
+	}
+
+	@Override
+	protected boolean renderPrimaryTableReference(TableGroup tableGroup, LockMode lockMode) {
+		final TableReference tableRef = tableGroup.getPrimaryTableReference();
+		// The H2 parser can't handle a sub-query as first element in a nested join
+		// i.e. `join ( (select ...) alias join ... )`, so we have to introduce a dummy table reference
+		if ( tableRef instanceof QueryPartTableReference || tableRef.getTableId().startsWith( "(select" ) ) {
+			final boolean realTableGroup = tableGroup.isRealTableGroup()
+					&& ( CollectionHelper.isNotEmpty( tableGroup.getTableReferenceJoins() )
+					|| hasNestedTableGroupsToRender( tableGroup.getNestedTableGroupJoins() ) );
+			if ( realTableGroup ) {
+				appendSql( "dual cross join " );
+			}
+		}
+		return super.renderPrimaryTableReference( tableGroup, lockMode );
+
+	}
+
+	@Override
+	protected boolean supportsRowValueConstructorSyntax() {
+		// Just a guess
+		return getDialect().getVersion().isSameOrAfter( 1, 4, 197 );
+	}
+
+	@Override
+	protected boolean supportsRowValueConstructorSyntaxInInList() {
+		// Just a guess
+		return getDialect().getVersion().isSameOrAfter( 1, 4, 197 );
+	}
+
+	@Override
+	protected boolean supportsRowValueConstructorSyntaxInQuantifiedPredicates() {
+		// Just a guess
+		return getDialect().getVersion().isSameOrAfter( 1, 4, 197 );
+	}
+
+	@Override
```

Comment:
Looks like there is something missing here from the current state of `H2Dialect`

```suggestion
	@Override
	protected boolean supportsNullPrecedence() {
		// Support for nulls clause in listagg was added in 2.0
		return getClauseStack().getCurrent() != Clause.WITHIN_GROUP || getDialect().getVersion().isSameOrAfter( 2 );
	}

	@Override
```